contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,672 @@
+
+        valInfo.jailed = false;
+
+        emit ValidatorUnjailed(validator);
+    }
+
+    function delegate(address validator) external payable onlyInitialized validatorExist(validator) validatorNotJailed(validator) whenNotPaused {
+        uint256 _bnbAmount = msg.value;
+        require(_bnbAmount >= minDelegationBNBChange, "INVALID_DELEGATION_AMOUNT");
+        address delegator = msg.sender;
+        Validator memory valInfo = validators[validator];
+        if (IStakePool(valInfo.poolModule).getSecurityDepositBNB() < minSelfDelegationBNB) {
+            // only self delegation
+            require(delegator == validator, "ONLY_SELF_DELEGATION");
+        }
+
+        emit Delegated(validator, delegator, _bnbAmount);
+
+        _updateEligibleValidators(UpdateDirection.Up, validator, valInfo.voteAddress);
+    }
+
+    function undelegate(address validator, uint256 _sharesAmount) public onlyInitialized validatorExist(validator) whenNotPaused {
+        require(_sharesAmount > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = validators[validator];
+        uint256 _bnbAmount = IStakePool(valInfo.poolModule).undelegate(delegator, _sharesAmount);
+
+        emit Undelegated(validator, delegator, _bnbAmount);
+
+        _updateEligibleValidators(UpdateDirection.Down, validator, valInfo.voteAddress);
+    }
+
+    function redelegate(address srcValidator, address dstValidator, uint256 _sharesAmount) public onlyInitialized validatorExist(srcValidator) validatorExist(dstValidator) validatorNotJailed(dstValidator) whenNotPaused {
+        require(_sharesAmount > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory srcValInfo = validators[srcValidator];
+        Validator memory dstValInfo = validators[dstValidator];
+
+        uint256 _bnbAmount = IStakePool(srcValInfo.poolModule).redelegate(delegator, _sharesAmount);
+        IStakePool(dstValInfo.poolModule).delegate{value: _bnbAmount}(delegator);
+
+        emit Redelegated(srcValidator, dstValidator, delegator, _bnbAmount);
+
+        bytes memory srcValVoteAddr = srcValInfo.voteAddress;
+        bytes memory dstValVoteAddr = dstValInfo.voteAddress;
+        _updateEligibleValidators(UpdateDirection.Down, srcValidator, srcValVoteAddr);
+        _updateEligibleValidators(UpdateDirection.Up, dstValidator, dstValVoteAddr);
+    }
+
+    function claimSecurityDeposit(uint256 _sharesAmount) external onlyInitialized validatorExist(msg.sender) {
+        Validator memory valInfo = validators[msg.sender];
+        address pool = valInfo.poolModule;
+        require(IStakePool(pool).balanceOf(address(this)) >= _sharesAmount, "NOT_ENOUGH_SHARES");
+
+        emit SecurityDepositClaimed(msg.sender, _sharesAmount);
+
+        IStakePool(pool).transfer(msg.sender, _sharesAmount);
+        if (IStakePool(pool).getSecurityDepositBNB() < minSelfDelegationBNB) {
+            valInfo.jailed = true;
```

Comment:
The validator should be jail for more than one day before validator can get his funds back, otherwise a validator can get funds back first and then do bad things.

Reply:
Done